Update README (#18)
inspec/train-habitat: README.md

# Train-Habitat

`train-habitat` is a Train plugin and is used as a Train Transport to connect to Habitat installations.

## To Install this as a User

You will need InSpec v2.3 or later.

Simply run:

```
$ inspec plugin install train-habitat
```

## Using train-habitat from InSpec

> As train-habitat takes potentially many options, it is simplest to list the options in your `~/.inspec/config.json` file, then used the named set of options with `-t`.

Reviewer: ```suggestion
As `train-habitat` takes potentially many options, it is simplest to list the options in your `~/.inspec/config.json` file, then used the named set of options with `-t`.
```

For example, if your config file contains:

```
{
  "file_version": "1.1",
  "credentials": {
    "habitat": {
      "dev-hab": {
        "api_url": "http://dev-hab.my-corp.io",
        "cli_ssh_host": "dev-hab.my-corp.io"
      },
      "prod-hab": {
        "api_url": "https://prod-hab.my-corp.io",
        "api_auth_token": "opensesame"
      },
    }
  }
}
```

Using this configuration, you could execute:

```
$ inspec exec some-profile -t habitat://dev-hab
# Or
$ inspec exec some-profile -t habitat://prod-hab
```

You may also pass `--config some-file.json` to use a config file at a different location.

> See below for the full list of options you may use with a `habitat` credential set in your configuration.

Reviewer: Is there a particular section that "See below" refers to? Would it be possible to use a jump link for that?

Author: Yes; I'm always a bit shy when it comes to internal anchor links, because I often guess wrong about how the page renderer handles spaces in the anchor (assuming it inserts anchors).  I'll experiment.

Reviewer: Totally understand. I know for .md that the internal link would be formatted as "[see below] (#header-name)" but I agree. I'm not quite sure how the page renderer treats such notation.

## Using train-habitat from Ruby

> The options that may be passed to `Train.create` are listed below.

Reviewer: Is there a particular section that "listed below" refers to? Would it be possible to use a jump link for that?

Author: Perhaps I should remove this one.  Here, 'below' means 'immediately following'.

Reviewer: D'oh. I finally noticed the header format for this section. I believe it would be okay to leave this sentence as is because the referenced content is immediately afterwards. I thought it was referencing content elsewhere in the document as with my previous comment about jump links.

### Dual-mode transport

> Because habitat exposes some facts by its HTTP Gateway API, and some facts by its CLI tool `hab`, this Train Transport has three modes of operation:

Reviewer: I was going to say that "habitat" should be `habitat`, but in reality, I think Habitat should be capitalized here. I can make that change for you if you'd like!

> Because habitat exposes some facts by its HTTP Gateway API, and some facts by its CLI tool `hab`, this Train Transport has three modes of operation:

Reviewer: ```suggestion
Because Habitat exposes some facts by its HTTP Gateway API, and some facts by its CLI tool `hab`, this Train Transport has three modes of operation:
```

 * Using only the HTTP API (no ability to query packages, but rich ability to query rings)
 * Using only the `hab` CLI command (limitations TBD)
 * Using both (full capabilities)

> When creating a train-habitat Connection, there are thus two sets of options, prefixed with `api_` and `cli_` respectively. You must provide at least one set.

Reviewer: ```suggestion
When creating a `train-habitat` Connection, there are thus two sets of options, prefixed with `api_` and `cli_` respectively. You must provide at least one set.
```

### API-Mode options

API-mode options are used to connect to a Habitat Supervisor running with an exposed HTTP Gateway. They are prefixed with `api_`.

```ruby
Train.create(:habitat, api_url: 'http://my-hab.my-company.io:9631')
```

#### api_url

> Required for API-mode use. An HTTP or HTTPS URL which identifies a Supervisor HTTP Gateway.  If the port is omitted from the URL, the API standard port of 9631 is assumed; to use port 80, specify it explicitly.

Reviewer: ```suggestion
Required for API-mode use. This is an HTTP or HTTPS URL that identifies a Supervisor HTTP Gateway.  If the port is omitted from the URL, the API standard port of 9631 is assumed; to use port 80, specify it explicitly.
```

#### api_auth_token

The supervisor may be configured to require a [Bearer Token Authorization](https://www.habitat.sh/docs/using-habitat/#monitor-services-through-the-http-api), in which the client and the gateway use a pre-shared secret. Use this option to specify the secret.

### CLI Mode options

> CLI options are more varied, and are entirely dependent on the underlying transport chosen to reach the CLI. For example, if there were a supported transport named 'radio' which took options 'channel' and 'band', specify them to train-habitat like this:

Reviewer: ```suggestion
CLI options are more varied, and are entirely dependent on the underlying transport chosen to reach the CLI. For example, if there were a supported transport named 'radio' that took options 'channel' and 'band', specify them to `train-habitat` like this:
```

```ruby
Train.create(:habitat, {cli_radio_band: 'VHF', cli_radio_channel: 23})
```

`train-habitat` identifies the underlying "sub-transport" using the prefixes of the provided options. For example, if you pass an option named `cli_ssh_host`, `train-habitat` will recognize that you intend to use the SSH transport to connect to a location that has access to the `hab` CLI tool.

You may specify many options referring to the same sub-transport (such as credentials), but it is an error to specify more than one CLI sub-transport.

Currently supported CLI transports include:
 * SSH

Plans for future support include (in approximate order):
 * WinRM
 * Local
 * Docker

#### General Options

> Any options not prefixed with `cli_` or `api_` are also passed to the CLI transport. This means you can use generic Train connection options such as the `sudo` and `shell` sets of options (see [train source code](https://github.com/inspec/train/blob/71679307903fc8853e09abd93f3901c83800e019/lib/train/extras/command_wrapper.rb#L31).), as well as `logger`.

Reviewer: ```suggestion
Any options not prefixed with `cli_` or `api_` are also passed to the CLI transport. This means you can use generic Train connection options such as the `sudo` and `shell` sets of options (see [train source code](https://github.com/inspec/train/blob/71679307903fc8853e09abd93f3901c83800e019/lib/train/extras/command_wrapper.rb#L31), as well as `logger`).
```

#### SSH Options

> `train-habitat` can accept any option that the Train SSH Transport accepts, if the prefix `cli_ssh_` is added. This includes:

Reviewer: ```suggestion
`train-habitat` can accept any option that the Train SSH Transport accepts if the prefix `cli_ssh_` is added. This includes:
```

$ inspec plugin install train-habitat

> ## Using the plugin as a Train Transport

Reviewer: Train Transport is used in this section, but not used elsewhere. Is this supposed to reference documentation in another repo?

Reviewer: Alternatively, defining "Train Transport" may be helpful in understanding context.

Author: Train is a system for communicating with remote systems.  Each type of remote system requires a different Transport.  Some Transports ship with train, and some (like this one) ship as plugins.

So, this whole project is a train plugin; one uses it as a Train Trainsport to connect to habitat.

Reviewer: So, to clarify, would the below work for amending line 3's description and provide that context to the rest of the document?
"train-habitat is a Train plugin and is used as a Train Transport to connect to Habitat installations."

### Dual-mode transport

Because habitat exposes some facts by its HTTP Gateway API, and some facts by its CLI tool `hab`, this Train Transport has three modes of operation:

 * Using only the HTTP API (no ability to query packages, but rich ability to query rings)
 * Using only the `hab` CLI command (limitations TBD)
 * Using both (full capabilities)

When creating a train-habitat Connection, there are thus two sets of options, prefixed with `api_` and `cli_` respectively. You must provide at least one set.

### API-Mode options

> 

Reviewer: Is there supposed to be more context on how to configure the API-Mode options? I'm comparing this section with CLI Mode options.

Author: Shall I add examples?

Reviewer: That would be wonderful! Another option is an overview explaining the API-Mode options under the API-Mode options header.

#### api_url

An HTTP or HTTPS URL which identifies a Supervisor HTTP Gateway.  If the port is omitted from the URL, the API standard port of 9631 is assumed; to use port 80, specify it explicitly.

#### api_auth_token

> The supervisor may be configured to require [Bearer Token Authorization](https://www.habitat.sh/docs/using-habitat/#monitor-services-through-the-http-api), in which the client and the gateway use a pre-shared secret. Specify the secret here to use this feature.

Reviewer: Where is "here" when specifying the secret? (last sentence)

Author: How about 'Use this option to specify the secret'?

Reviewer: Prefect.
